### ISSUE TITLE
Filestore plugins

### ISSUE THREAD

**Issue by magik6k** [ipfs/go-ipfs#5581]
It would be really nice to be able to extend filestore with custom storage mechanisms.

The use case I have in mind is referencing git objects in packed repositories. Currently the problem is that storing raw git objects is hugely inefficient in terms of storage, for example the linux-stable repo is about 1.5G, unpacked/added to ipfs it becomes 40G+

cc @kevina

**Comment by Kubuxu** [comment]
We could think about adding a fast compression layer as a part of datastore composition.

**Comment by Stebalien** [comment]
Yeah, we should definitely have storage-layer compression. However, we'll probably want a links-service first.
